system/acpi.c

```diff
@@ -0,0 +1,345 @@
+        rp = scan_for_rsdp(0xE0000, 0x20000);
+        if (rp) rsdp_source = "BIOS reserved area";
+    }
+    if (rp == NULL) {
+        // RSDP not found, give up.
+        return 0;
+    }
+    return (uintptr_t)rp;
+}
+
+static uintptr_t find_acpi_table(uint32_t table_signature)
+{
+   rsdp_t *rp = (rsdp_t *)acpi_config.rsdp_addr;
+
+    // Found the RSDP, now get either the RSDT or XSDT
+    // and scan it for a pointer to the table we're looking for
+    rsdt_header_t *rt;
+
+    if (acpi_config.ver_maj < rp->revision) {
+        acpi_config.ver_maj = rp->revision;
+    }
+
+    if (rp->revision >= 2) {
+        rt = (rsdt_header_t *)map_region(rp->xsdt_addr, sizeof(rsdt_header_t), true);
+        if (rt == NULL) {
+            return 0;
+        }
+        // Validate the XSDT.
+        if (*(uint32_t *)rt != XSDTSignature) {
+            return 0;
+        }
+        rt = (rsdt_header_t *)map_region(rp->xsdt_addr, rt->length, true);
+        if (rt == NULL || acpi_checksum(rt, rt->length) != 0) {
+            return 0;
+        }
+        // Scan the XSDT for a pointer to the table we're looking for.
+        uint64_t *tab_ptr = (uint64_t *)((uint8_t *)rt + sizeof(rsdt_header_t));
+        uint64_t *tab_end = (uint64_t *)((uint8_t *)rt + rt->length);
+
+        while (tab_ptr < tab_end) {
+            uintptr_t addr = *tab_ptr++;  // read the next table entry
+            uint32_t *ptr = (uint32_t *)map_region(addr, sizeof(uint32_t), true);
+
+            if (ptr && *ptr == table_signature) {
+                return addr;
+            }
+        }
+    } else {
+        rt = (rsdt_header_t *)map_region(rp->rsdt_addr, sizeof(rsdt_header_t), true);
+        if (rt == NULL) {
+            return 0;
+        }
+        // Validate the RSDT.
+        if (*(uint32_t *)rt != RSDTSignature) {
+            return 0;
+        }
+        rt = (rsdt_header_t *)map_region(rp->rsdt_addr, rt->length, true);
+        if (rt == NULL || acpi_checksum(rt, rt->length) != 0) {
+            return 0;
+        }
+        // Scan the RSDT for a pointer to the table we're looking for.
+        uint32_t *tab_ptr = (uint32_t *)((uint8_t *)rt + sizeof(rsdt_header_t));
+        uint32_t *tab_end = (uint32_t *)((uint8_t *)rt + rt->length);
+
+        while (tab_ptr < tab_end) {
+            uintptr_t addr = *tab_ptr++;  // read the next table entry
+            uint32_t *ptr = (uint32_t *)map_region(addr, sizeof(uint32_t), true);
+
+            if (ptr && *ptr == table_signature) {
+                return addr;
+            }
+        }
+    }
+
+    return 0;
+}
+
+static bool parse_fadt(uintptr_t fadt_addr)
+{
+    // FADT is a very big & complex table and we only need a few data.
+    // We use byte offset instead of a complete struct.
+
+    // FADT Header is identical to RSDP Header
+    rsdt_header_t *fadt = (rsdt_header_t *)fadt_addr;
+
+    // Validate FADT
+    if (fadt == NULL || acpi_checksum(fadt, fadt->length) != 0) {
```

Comment:
Ditto about fadt->length.

Reply:
Still according to ACPI specs : https://uefi.org/specs/ACPI/6.4/05_ACPI_Software_Programming_Model/ACPI_Software_Programming_Model.html#fixed-acpi-description-table-fadt

Reply:
Well, when standards do not define safeguards, we'd better invent ours, as a measure against voluntarily or involuntarily specially crafted data, whose processing would consume excessive CPU time, or cause memory accesses beyond the currently mapped memory in some cases :)
In the absence of such checks, fuzzers ought to quickly destroy the parsing code...